src/System.Management.Automation/engine/CoreAdapter.cs

```diff
@@ -2788,6 +2788,14 @@ internal PropertyCacheEntry(PropertyInfo property)
                 if (propertySetter != null && (propertySetter.IsPublic || propertySetter.IsFamily))
                 {
                     this.isStatic = propertySetter.IsStatic;
+                    foreach (Type requiredCustomModifier in propertySetter.ReturnParameter.GetRequiredCustomModifiers())
```

Comment:
@IISResetMe Sorry for taking so long. The changes here doesn't prevent `PSProperty` from being changed for an `init` property. Example (with changes in this PR):
```
> Add-Type -TypeDefinition @'
    public class InitOnlyTestClass
    {
        // default constructor to allow cast-initialization
        public InitOnlyTestClass()
        {
        }
    
        public long Value { get; init; }
    }
'@

> $s = [InitOnlyTestClass]::new()
> $p = $s.psobject.Properties['Value']
> $p.IsSettable
False
> $p.Value = 1000
> $p

MemberType      : Property
Value           : 1000
IsSettable      : False
IsGettable      : True
TypeNameOfValue : System.Int64
Name            : Value
IsInstance      : True
```

-----------------------------------------

The hashtable-to-object conversion actually goes through [ConvertViaNoArgumentConstructor.Convert](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/LanguagePrimitives.cs#L3970), and depends on [setting `PSProperty.Value`](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/LanguagePrimitives.cs#L4701).
This is the tricky part, for an `init` property, `PSProperty.Value` should throw when setting value to it, but the `[type]@{ 
property1 = value1; property2 = value2 }` depends on setting `PSProperty.Value`, so we got a contradiction here.

Reply:
I'm almost inclined to say that's a restriction we could document. Going through PSObject to target the underlying metadata and modify the PSProperty directly is almost akin to using Reflection to modify the value in C# -- and in Reflection, too, I don't think init-only holds any water.

Reply:
I personally don't think we should introduce a language feature that is inconsistent. If a user is doing reflection directly, then it's out of our hands. But if the user is doing it in the PowerShell way, then consistent behavior should be expected.

Reply:
@daxian-dbw I agree behavior should be consistent - but I'm inclined to argue that inconsistent _is consistent_ with the behavior we're trying to mimic here. 

Init-only properties _are an illusion_, in that it requires the consenting participation of all compilers. The C#, VB and F# implementations all introduce compile-time guardrails, no actual runtime-constraints or CTS-specific flags to indicate any behaviorial difference between properties defined with `init;` and auto properties. Because there is no such flags. The core design of CTS anticipates no such distinction :) 

We can mimic the runtime-inconsistent behavior that the dominant .NET languages have converged on (which is what this PR attempts), or we can decide to adapt differently (because PowerShell is different and its runtime has a different lifecycle model).

Perhaps I should close this PR and leave that question up to the language WG? 😄